### PR TITLE
[Draft] Flexible table creation

### PR DESCRIPTION
**Pull request by jjl**
Will solve #238. Haven't yet written tests or done the non-pg adapters, but does this approach look ok?

### CODE REVIEW

File: lib/ecto/adapters/postgres/connection.ex [hunk 1]
```diff
@@ -789,6 +789,10 @@ if Code.ensure_loaded?(Postgrex) do
       ?t
     end
 
+    defp table_type_expr(nil), do: "TABLE "
+    defp table_type_expr(type) when is_binary(type), do: type <> " "
+    defp table_type_expr(type), do: raise ArgumentError, message: "Table type must be a String, got " <> inspect(type)
```
Comment (josevalim) [inline lib/ecto/adapters/postgres/connection.ex:794]: Just for consistency with other error messages:

```suggestion
    defp table_type_expr(type), do: raise ArgumentError, "table type must be a string, got `#{inspect(type)}`"
```

**Comment by josevalim** [comment]
Yup, this is exactly the way to go! You may need to change alter and drop table statements too.

**Comment by jjl** [comment]
ALTER turns out to be a bit more complicated, specifically altering to add a constraint, because there is no table object available to take the type from. Not sure how to resolve this, please advise.

**Comment by josevalim** [comment]
Which alter? Our command should be alter table(...), no?

**Comment by jjl** [comment]
[this one](https://github.com/elixir-ecto/ecto_sql/blob/master/lib/ecto/adapters/postgres/connection.ex#L875)

**Comment by josevalim** [comment]
Let's add a `type` field to the constraint too. WDYT?

**Comment by jjl** [comment]
it would be table_type, and my gut feeling is that it belongs in a table (and thus i was wondering about passing a table object in). But I'm not entirely clear how we end up with one of these Constraint objects.

**Comment by josevalim** [comment]
The API is `create index(:users, [:foo])`. I guess one option it to allow `create index(table(:foo), [...])`, maybe even rewrite the atom to the table, but it sounds a bit too much. Any other thoughts?

**Comment by jjl** [comment]
That doesn't seem too bad a solution to be honest.

**Comment by josevalim** [comment]
So let's do that. if an atom is given, we automatically wrap it in a table. Otherwise a table has to be given. If you prefer, you can start on this change first, before this PR, and then once that is done, take back this PR. But I assume that is not such a large change anyway, so altogether would also be fine.

**Comment by josevalim** [comment]
Ping. We plan to release Ecto v3.5 in a week or two, so if we want to get this in, now is the time. Otherwise it will probably be in a couple months. :)

**Comment by josevalim** [comment]
Ping.

**Comment by josevalim** [comment]
I will go ahead and close this for now. If there is still an interest in getting this in, please let me know and I will gladly reopen!

**Comment by jjl** [comment]
Sorry Jose, I've been spending all my time with rust recently and my elixir work has barely had a look in. I do intend to land this at some point, it's just my brain is elsewhere.

**Pull request closed without merging**